<!DOCTYPE html>
<!--
    Josh Hong
    howtouse.html
    Jun 2, 2022
-->
<html lang="en">
<head>
    <title>DMEA: How to Use</title>
    <meta charset="utf-8">
    <link rel="stylesheet" href="css/styles.css">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&display=swap" rel="stylesheet">
    <link rel="shortcut icon" href="images/favicon.ico">
    <link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png">
    <link rel="apple-touch-icon" sizes="180x180" href="images/apple-touch-icon.png">
    <link rel="icon" sizes="192x192" href="images/android-chrome-192x192.png">
</head>
<body>
    <div id="wrapper">
        <header>
                <a href="index.html"><h1>DMEA</h1></a>
                <h2>Drug Mechanism Enrichment Analysis</h2>
        </header>
    
        <nav>
            <ul>
                <li><a href="index.html">Home</a></li>
                <li><a href="howtouse.html">How to Use</a></li>
                <li><a href="why.html">Why Use DMEA</a></li>
                <li><a href="methods.html">Methods</a></li>
                <li><a href="about.html">About Us</a></li>
                <li><a href="contact.html">Contact</a></li>
            </ul>
        </nav>
    
        <!-- Use the main area to add the main content to the webpage -->
        <main>

            <div>
                <section id="how">
                    <h2>How to Use: Web App</h2>
                    <p>The easiest way to use DMEA is with the web application here: <a href="https://belindabgarana.shinyapps.io/dmea" class="link" target="_blank">https://belindabgarana.shinyapps.io/dmea</a> . For more information, please watch our demo video below or <a href="contact.html" class="link">Contact Us</a>. </p>
                    <iframe width="560" height="315" src="https://www.youtube.com/embed/Kt0Bq6coCdA" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
                    
                    <p>To use the web app, you can either run one of our provided examples or select one of 4 input types:</p>

                    <ol>
                        <li>CMap L1000 query output .gct file (i.e., “query_result.gct” or “ncs.gct”)</li>
                        <li>CMap PRISM query output .gct file (i.e., “ncs.gct”)</li>
                        <li>Drug rank list .csv file (drug names in column 1, drug rank metrics in column 2, optional: drug mechanisms-of-action in column 3; headers included)</li>
                        <li>Gene signature .csv file (gene names in column 1, gene rank metrics in column 2; headers included)</li>
                    </ol>

                    <p>
                        DMEA is flexible to accept any signed numeric values as rank metrics for drugs or genes.
                        Please refer to the example inputs below or the “Examples” folder on our GitHub repo: <a href="https://github.com/BelindaBGarana/DMEA" class="link" target="_blank">https://github.com/BelindaBGarana/DMEA</a>
                    </p>
                </section>

                <section id="rank">
                    <h3>Drug Rank List</h3>
                    <img src="images/Drug_rank_list_format.png" alt="drug rank list format">

                    <p>If the third column with moa annotations is not provided, our default moa annotations based on the PRISM drug screen dataset will be used.</p>
                </section>

                <section id="signature">
                    <h3>Gene Signature</h3>
                    <img src="images/Gene_signature_format.jpg" alt="gene signature format">

                    <p>Gene signature inputs will be processed using the PRISM drug sensitivity score metric AUC (Area Under the Curve) and CCLE RNAseq data from version 19Q4.</p>
                </section>

            </div>

            <section id="how">
                <h2>How to Use: R Package</h2>
                <p>
                    To use the DMEA R package, first install it by running the 2 lines of code below 
                    and then use the steps below to either input a drug rank list or a gene signature.
                    If you use Windows OS, you may need to install the R package using 
                    devtools::install_github ('BelindaBGarana/DMEA', build=FALSE).
                    DMEA is flexible to accept any signed numeric values as rank metrics for drugs or genes. 
                    To see examples, please refer to the “Examples” folder on our GitHub repo here: <a href="https://github.com/BelindaBGarana/DMEA" class="link" target="_blank">https://github.com/BelindaBGarana/DMEA</a>
                </p>
                <div id="info" class="round">
                    <p class="equation">if (!require(devtools)) {install.packages("devtools")}</p>
                    <p class="equation">devtools::install_github ('BelindaBGarana/DMEA')</p>
                </div>
            </section>

            <section id="rank">
                <h3>Drug Rank List</h3>
                <ol>
                    <li>Install the DMEA R package and load it from your library</li>
                    <li>Input a dataframe containing your drug rank list. 
                        If you have mechanism-of-action (MOA) annotations, you can provide these in a column in your dataframe. 
                        If not, you must also input a gmt object such as “MOA_gmt_file_n6_no_special_chars.gmt” available in the “Inputs” folder on our GitHub repo. 
                        2+ MOA must be represented in your input.</li>
                    <li>Save the output dataframe of the enrichment results for each drug set in addition to plots summarizing the results</li>
                </ol>
                <img src="images/Fig1_DMEA_2022-06-24_no_background.png" alt="DMEA method for drug rank list input">
            </section>

            <section id="signature">
                <h3>Gene Signature</h3>
                <ol>
                    <li>Install the DMEA R package and load it from your library</li>
                    <li>Input a dataframe containing your gene signature, as well as dataframes containing drug sensitivity scores and gene expression data with common sample names in the first columns. 
                        If you have mechanism-of-action (MOA) annotations, you can provide these in a separate dataframe. 
                        If not, you must also input a gmt object such as “MOA_gmt_file_n6_no_special_chars.gmt”. 
                        This file, in addition to files containing drug sensitivity scores and gene expression data, is available in the "Inputs" folder on our GitHub repo. 
                        2+ MOA must be represented in your input.</li>
                    <li>Save the output dataframes with the Weighted Voting, correlation, and enrichment results for each drug set in addition to plots summarizing the results</li>
                </ol>
                <div id="sigimage">
                    <img src="images/Fig4_DMEA_1000dpi.png" alt="DMEA method for gene signature input">
                </div>
            </section>
        </main>
    
        <footer>
            <p>&copy; Copyright 2022-2023. All Rights Reserved</p>
            <p><a href="mailto:nagraham@usc.edu">nagraham@usc.edu</a></p>
        </footer>
    </div>
</body>
</html>
