---
title: "Drug Mechanism Enrichment Analysis"
author: 
  - name: Belinda B. Garana
  - name: James H. Joly
package: DMEA
output:
  BiocStyle::html_document:
  vignette: >
    %\VignetteIndexEntry{Drug Mechanism Enrichment Analysis}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteIEncoding{UTF-8}
      toc_float: true
      toc_depth: 2
---

# Introduction
Though there are many algorithms for drug repurposing, 
most only evaluate individual drugs which can be prone to off-target effects. 
In contrast, our approach is to evaluate drug mechanisms of action by 
aggregating results across many drugs. Though mechanisms of action have also 
been evaluated in a similar manner by the Connectivity Map (clue.io), 
Drugmonizome (https://maayanlab.cloud/drugmonizome), and DrugEnrichr 
(https://maayanlab.cloud/DrugEnrichr), these tools only allow analysis of select 
public data sets and do not consider the ranks or direction of input features. 
With our tool Drug Mechanism Enrichment Analysis (DMEA), users can query any 
dataset of interest using either an input gene signature or drug rank list to 
identify enriched mechanisms of action for drug repurposing efforts. For more 
information, please visit our website: https://belindabgarana.github.io/DMEA

# Citation
To cite this package, please use:
Garana, B.B., Joly, J.H., Delfarah, A. et al. Drug mechanism enrichment analysis improves prioritization of therapeutics for repurposing. BMC Bioinformatics 24, 215 (2023). https://doi.org/10.1186/s12859-023-05343-8

# Installation
To install this package, start R and enter:
```{r eval=FALSE}
if (!require("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

BiocManager::install("DMEA")
```

Alternatively, to install this package directly from GitHub, start R and enter:
```{r eval=FALSE}
if (!require("devtools", quietly = TRUE)) {
  install.packages("devtools")
}

devtools::install_github("BelindaBGarana/DMEA")
```

If you are using Windows OS, you may need to change the code above to:
```{r eval=FALSE}
if (!require("devtools", quietly = TRUE)) {
  install.packages("devtools")
}

devtools::install_github("BelindaBGarana/DMEA", build = FALSE)
```

# Input Option 1: Drug Rank List
With our drugSEA function, you can calculate the enrichment of drug sets in 
an input drug rank list. This drug rank list can be derived from anywhere 
(e.g., various drug repurposing algorithms). The only requirements are that:
  * Each drug has only one numeric, nonzero rank metric
  * At least 2 drug sets are represented in your list

Required inputs:
  * data: a data frame containing one column for drug names and another column 
          for drug ranks
  * gmt: gmt object containing drug set information. If not provided, a third 
         column in your data input parameter must include set annotations; if 
         so, the name of this column must be denoted by the set.type parameter. 
         If you do not have your own drug set annotations, you can use our 
         PRISM moa annotations on our GitHub repository at: https://raw.github.com/BelindaBGarana/DMEA/shiny-app/Inputs/MOA_gmt_file_n6_no_special_chars.gmt

Example:
1. Prepare data frame containing rank metrics and set annotations for each drug
```{r}
# Drug is our default name for the column containing drug names
Drug <- paste0("Drug_", seq(from = 1, to = 24))

data <- as.data.frame(Drug)

# Pearson.est is our default name for the column containing drug ranks
data$Pearson.est <- rnorm(length(Drug), sd = 0.25)

# moa is our default name for the column containing drug set annotations
data$moa <- rep(paste("Set", LETTERS[seq(from = 1, to = 4)]), 6)
```

2. Perform drugSEA and store results
```{r}
# if you use other column names for your data parameter,
# then make sure to set drug, rank.metric, and set.type parameters
drugSEA.test <- DMEA::drugSEA(data)
```

# Input Option 2: Gene Signature
With our DMEA function, you can identify selectively toxic drug sets based on 
the correlations between drug sensitivity scores (e.g., AUC) and weighted gene 
voting scores generated with an input gene signature. This requires having 
drug screen and expression datasets which share the same samples (e.g., cell 
lines). To run this analysis, it is required that:
  * Each drug has only one numeric, nonzero correlation metric
  * At least 2 drug sets are represented in your drug screen data
  * Each gene has only one numeric weight value
  * At least one gene symbol matches between the input gene signature and 
    expression data frames
  * At least 3 samples (e.g., cell lines) are in both the expression and drug 
    screen sensitivity score data frames

Required inputs:
  * drug.sensitivity: a data frame containing drug sensitivity scores (e.g., 
    AUC) with drug names as column names, except for one column which contains 
    sample names
  * gmt: gmt object containing drug set information. If not provided, a 
    drug.info parameter with set membership information must be provided. If 
    you do not have your own drug set annotations, you can use our PRISM moa 
    annotations on our GitHub repository at: https://raw.github.com/BelindaBGarana/DMEA/shiny-app/Inputs/MOA_gmt_file_n6_no_special_chars.gmt
  * expression: a data frame containing normalize gene expression with gene 
    symbols as column names, except for one column which contains sample names
  * weights: a data frame containing gene symbols in one column (default: 
    column 1) and weight values in another column (default: column 2)

Example:
1. Prepare drug sensitivity data frame
```{r}
# create list of sample names
Sample_ID <- seq(from = 1, to = 21)

drug.sensitivity <- as.data.frame(Sample_ID)

# create list of drug names
Drug <- paste0("Drug_", seq(from = 1, to = 24))

# give each drug values representative of AUC sensitivity scores
for (i in 1:length(Drug)) {
  drug.sensitivity[, c(Drug[i])] <- rnorm(length(Sample_ID),
    mean = 0.83,
    sd = 0.166
  )
}
```

2. Prepare drug info data frame
```{r}
# alternatively, a gmt object could be provided
info <- as.data.frame(Drug)

# moa is our default name for the column containing drug set annotations
info$moa <- rep(paste("Set", LETTERS[seq(from = 1, to = 4)]), 6)
```

3. Prepare gene weight data frame
```{r}
# create list of gene symbols
Gene <- paste0("Gene_", seq(from = 1, to = 50))

# by default, gene symbols are found in
# the first column of your weights data frame
weights <- as.data.frame(Gene)

# give each gene a weight
# by default, gene weight values are found in
# the second column of your weights data frame
weights$Rank_metric <- rnorm(length(Gene))
```

4. Prepare expression data frame
```{r}
# by default, column 1 of your expression data frame is
# the column name from which sample names are gathered and
# the column containing sample names in your drug sensitivity
# data frame should have the same name
expr <- as.data.frame(Sample_ID)

# give each gene values representative of normalized RNA expression
# each gene is represented by a column in your expression data frame
for (i in 1:length(Gene)) {
  expr[, c(Gene[i])] <- rnorm(length(Sample_ID), sd = 0.5)
}
```

5. Perform DMEA and store results
```{r}
DMEA.test <- DMEA::DMEA(drug.sensitivity,
  expression = expr,
  weights = weights, drug.info = info
)
```

# Session Info
```{r}
sessionInfo()
```
